tests/cbmc/proofs/s2n_hash_const_time_get_currently_in_hash_block/s2n_hash_const_time_get_currently_in_hash_block_harness.c

```diff
@@ -0,0 +1,43 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+#include "api/s2n.h"
+#include "crypto/s2n_hash.h"
+
+#include <cbmc_proof/make_common_datastructures.h>
+#include <cbmc_proof/proof_allocators.h>
+
+void s2n_hash_const_time_get_currently_in_hash_block_harness()
+{
+    /* Non-deterministic inputs. */
+    struct s2n_hash_state *state = cbmc_allocate_s2n_hash_state();
+    uint64_t size;
+    uint64_t* out = bounded_malloc(size);
+
+    /* Assumptions. */
+    __CPROVER_assume(s2n_hash_state_is_valid(state));
+    if (state != NULL)
+    {
+        __CPROVER_file_local_s2n_hash_c_s2n_hash_set_impl(state);
+    }
```

Comment:
https://github.com/awslabs/s2n/blob/main/crypto/s2n_hash.c#L636-L638 Do we want to add this as an assumption? Or enforce that it's true in some other way?

Reply:
We could enforce this by doing something similar to other, for instance,
```c
int s2n_hash_reset(struct s2n_hash_state *state)
{
    notnull_check(state);
    /* Ensure that hash_impl is set, as it may have been reset for s2n_hash_state on s2n_connection_wipe.
     * When in FIPS mode, the EVP API's must be used for hashes.
     */
    GUARD(s2n_hash_set_impl(state));
```
What do you suggest? Should I add `GUARD(s2n_hash_set_impl(state));` to `s2n_hash_const_time_get_currently_in_hash_block` function?

Reply:
My concern is with the "Requires that hash_block_size is a power of 2" part of the comment, not whether a hash implementation exists.

Reply:
@lrstewart we get `hash_block_size` using `s2n_hash_block_size`,
```c
    GUARD(s2n_hash_block_size(state->alg, &hash_block_size));
```

which always returns a value that is power of two.

```c
int s2n_hash_block_size(s2n_hash_algorithm alg, uint64_t *block_size)
{
    PRECONDITION_POSIX(S2N_MEM_IS_READABLE(block_size, sizeof(*block_size)));
    switch(alg) {
    case S2N_HASH_NONE:       *block_size = 64;   break;
    case S2N_HASH_MD5:        *block_size = 64;   break;
    case S2N_HASH_SHA1:       *block_size = 64;   break;
    case S2N_HASH_SHA224:     *block_size = 64;   break;
    case S2N_HASH_SHA256:     *block_size = 64;   break;
    case S2N_HASH_SHA384:     *block_size = 128;  break;
    case S2N_HASH_SHA512:     *block_size = 128;  break;
    case S2N_HASH_MD5_SHA1:   *block_size = 64;   break;
    default:
        S2N_ERROR(S2N_ERR_HASH_INVALID_ALGORITHM);
    }
    return S2N_SUCCESS;
}
```

Do you think it's worth to add an extra check? For example,
```c
    uint64_t hash_block_size;
    GUARD(s2n_hash_block_size(state->alg, &hash_block_size));
    /**
     * Requires that hash_block_size is a power of 2. This is true for all hashes we currently support
     * If this ever becomes untrue, this would require fixing this.
     */
    ENSURE_POSIX(x && (!(x & (x-1))));
    *out = state->currently_in_hash & (hash_block_size - 1);
```

Reply:
Eh I don't think this needs to be a runtime check, since it is based on a small set of implementations. Maybe we can add the check to the s2n_hash_block_size CBMC proof?

Reply:
Implemented as suggested. @lrstewart could you take another look?